### PR TITLE
Safe unnesting of Config and Injector

### PR DESCRIPTION
**Pull request by dhensby**
Currently this only seriously affects `Injector`.

Injector declares that the parameter to `set_inst` must be an `Injector` and so if you unnest too far, it will break because it attempts to set an injector inst of `null`.

Thoughts:

Perhaps if we unnest too far it should just instantiate a new inst explicitly, rather than returning the current inst... Though I don't feel an arbitary unnest should discard the route config, instead, a dev should explicitly do that if it's their desire to reset the config completely.

### CODE REVIEW

**Comment by micmania1** [comment]
So hard to read the modified files btw - too much whitespace changes.

Should we not just throw an error when Config::set_inst() is passed anything other than an instance of Config? I don't see benefit in trying to fix a users issue, at least not without alerting them that something is wrong.

**Comment by tractorcow** [comment]
I think it should crash with a message instead of trying to silently
recover.

**Comment by dhensby** [comment]
@micmania1 sorry, you can see non-whitespace changes by appending `?w=1` to the files url (https://github.com/silverstripe/silverstripe-framework/pull/3982/files?w=1) but I'll stop with my whitespace cleaning in PRs.

I'm still quite torn between whether the framework should be lenient/forgiving or strict.

This way dev's can be a bit lazy and unnest more than they nest and they'll never lose the initial settings for the app (which I think is pretty important to retain). It's not infeasible to imagine a scenario where a complex set of nesting and unnesting is done where keeping track of nesting counts is difficult.

Throwing a logic exception "Can't unnest Config that is not nested" makes sense, but should it really break the system if it's recoverable? The debate would be how is best to recover if we think killing the application is not a good idea here.

Personally, I'm a fan of being strict in my own projects, but that's not really how the framework is in general and I don't think its conducive to encouraging our community to use SS. By being strict, we force our devs to write more code (either try catch or checking if the Config/Injector is nested).

If we threw errors, we'd need to provide an API to check if the config was currently nested and probably a way to access the root `Config`/`Injector`.

Perhaps throwing a warning and then retaining the root `Config`/`Injector` is  the best of both? We can push better code whilst allowing for sloppiness?

**Comment by dhensby** [comment]
@chillu / @hafriedlander / @sminnee - your thoughts on strictness would be appreciated here

**Comment by tractorcow** [comment]
Perhaps a crash on dev/test and silently ignored on live?

**Comment by dhensby** [comment]
Yes, that's how I'm leaning

File: core/Config.php [hunk 1]
```diff
@@ -231,13 +231,16 @@ public static function nest() {
...
 	 * Config object was copied from.
-	 * 
+	 *
 	 * @return Config Reference to new active Config instance
 	 */
 	public static function unnest() {
-		return self::set_instance(self::$instance->nestedFrom);
+		if (self::inst()->nestedFrom) {
+			self::set_instance(self::inst()->nestedFrom);
+		}
```
Comment (tractorcow) [inline core/Config.php]: Please use a warning here in an else so that we can catch these errors during dev.

**Pull request merged**